photosynthesis_metrics/fid.py

```diff
@@ -1,45 +1,39 @@
-import torch
+"""
+PyTorch implementation of Frechet Inception Distance (FID score)
+Reference:
+    Martin Heusel et al. "GANs Trained by a Two Time-Scale Update Rule Converge to a Local Nash Equilibrium",
+    https://arxiv.org/abs/1706.08500
+Credits:
+    https://github.com/hukkelas/pytorch-frechet-inception-distance/
+    https://github.com/mseitzer/pytorch-fid
+"""
+
+from typing import Callable, Tuple
 
 import numpy as np
-
+import torch
 from scipy import linalg
-from typing import Tuple, Optional
 from torch import nn
 
-from photosynthesis_metrics.feature_extractors.fid_inception import InceptionV3
-
-
-def frechet_inception_distance(predicted_images: torch.Tensor, target_images: torch.Tensor,
-                               feature_extractor: Optional[nn.Module] = None) -> float:
-    assert isinstance(predicted_images, torch.Tensor) and isinstance(target_images, torch.Tensor), \
-        f'Both image stacks must be of type torch.Tensor, got {type(predicted_images)} and {type(target_images)}.'
-
-    # There is no assert for the full equality of shapes because it is not required.
-    # Stacks can have any number of elements, but shapes of feature maps obtained from the images need to be equal.
-    # Otherwise it will not be possible to correctly compute statistics.
-    predicted_image_shape, target_image_shape = predicted_images.shape[1:], target_images.shape[1:]
-    assert predicted_image_shape == target_image_shape, \
-        f'Both image stacks must have images of the same shape, got {predicted_image_shape} and {target_image_shape}.'
-    assert isinstance(feature_extractor, nn.Module) or feature_extractor is None, \
-        f'Only PyTorch models are supported as feature extractors, got {type(feature_extractor)}.'
-
-    if feature_extractor is None:
-        print('WARNING: default feature extractor (InceptionNet V2) is used.')
-        feature_extractor = InceptionV3()
-
-    predicted_features = feature_extractor(predicted_images)
-    target_features = feature_extractor(target_images)
-
-    # TODO: `compute_fid` works with np.arrays, but need to work with torch.Tensors. Refactor that
-    return compute_fid(predicted_features, target_features)
-
+from photosynthesis_metrics.utils import BaseFeatureMetric, _validate_features
 
 def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray, eps=1e-6) -> float:
```

Comment:
Please change type annotations for `mu1` and `mu2` here. Now they are set to be float, but you alway pass results of `compute_statistics` here, which are `Tuple[np.ndarray, np.ndarray]`

```suggestion
def __compute_fid(mu1: np.ndarray, sigma1: np.ndarray, mu2: np.ndarray, sigma2: np.ndarray, eps=1e-6) -> float:
```